api/proto/teleport/dbobjectimportrule/v1/dbobjectimportrule.proto

```diff
@@ -0,0 +1,82 @@
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//      http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+syntax = "proto3";
+
+package teleport.dbobjectimportrule.v1;
+
+import "teleport/header/v1/metadata.proto";
+
+option go_package = "github.com/gravitational/teleport/api/gen/proto/go/teleport/dbobjectimportrule/v1;dbobjectimportrulev1";
+
+// DatabaseObjectImportRule is the resource representing a global database object import rule.
+message DatabaseObjectImportRule {
+  // The kind of resource represented.
+  string kind = 1;
+  // Differentiates variations of the same kind. All resources should
+  // contain one, even if it is never populated.
+  string sub_kind = 2;
+  // The version of the resource being represented.
+  string version = 3;
+  // Common metadata that all resources share.
+  teleport.header.v1.Metadata metadata = 4;
+
+  // spec represents the specifications for the database object import rule.
+  DatabaseObjectImportRuleSpec spec = 5;
+}
+
+// DatabaseObjectImportRuleSpec is the spec for database object import rule.
+message DatabaseObjectImportRuleSpec {
+  // priority represents the priority of the rule application. Rules with higher priorities override those with lower.
+  int32 priority = 1;
+
+  // db_labels is a set of labels matched against database labels.
+  map<string, string> db_labels = 2;
+
+  // Mappings is a list of matches that will map match conditions to labels.
+  repeated DatabaseObjectImportRuleMapping mappings = 4;
+}
+
+// DatabaseObjectImportRuleMapping is the mapping between object properties and labels that will be added to the object.
+message DatabaseObjectImportRuleMapping {
+  // Match specifies the matching rules.
+  DatabaseObjectImportMatch match = 1;
+
+  // Scope specifies the object scope.
+  DatabaseObjectImportScope scope = 2;
+
+  // AddLabels specifies which labels to add if the match succeeds.
+  map<string, string> add_labels = 3;
+}
+
+// DatabaseObjectImportMatch specifies the matching rules.
+message DatabaseObjectImportMatch {
+  // table_names specify the names of the tables to match.
+  repeated string table_names = 1;
+
+  // view_names specify the names of the views to match.
+  repeated string view_names = 2;
+
+  // procedure_names specify the names of the procedures to match.
+  repeated string procedure_names = 3;
+}
+
+// DatabaseObjectImportScope specifies the object scope.
+message DatabaseObjectImportScope {
+  // database_names specify the names of the databases to include in the scope.
+  repeated string database_names = 1;
```

Comment:
If there are multiple database_names and schema_names, what is the final match? The combination of all databases and schemas? Please document.

Is there a way to represent database+schema in a way that it wouldn't create a combinatory explosion? Or is the answer to that "do multiple mappings" ?

I guess what I'm trying to say is: is making these fields repeated a reasonable design choice, or should we change them to non-repeated and require multiple mappings instead?

Reply:
Please see the RFD for the details. Given that RFD is commited to the repo, I consider it part of the project documentation I can refer people too; I'll add a note to read it for details behind these data types.

Reply:
I've added a note to go to RFD for details.

Reply:
Please document the answer to the first question, it won't take much of your time and it will be helpful to everyone who is not yet in the know and is trying to make sense of this protos, like myself and other reviewers.

I would appreciate if you could answer the other two questions, even if just in this comment thread, or maybe point me to a specific part of the RFD that addresses this.

Reply:
> If there are multiple database_names and schema_names, what is the final match? The combination of all databases and schemas? Please document.

You don't care which particular entry in database_names did you match, same for schema_names. Overall, the result of a match is boolean.

> Is there a way to represent database+schema in a way that it wouldn't create a combinatory explosion? Or is the answer to that "do multiple mappings" ?

DatabaseObjectImportScope is isomorphic with `([]string, []string)` pair, rather than `[](string, string)`.We simply match against both fields independently, with wildcard `*` support.

Here is the actual code (planned; can't have it on master without merging its dependencies first):

https://github.com/gravitational/teleport/blob/c0afc5d8b642a4e68c4a5f99c96c52383f382ccf/lib/srv/db/common/permissions/import_rules.go#L99-L116

> I guess what I'm trying to say is: is making these fields repeated a reasonable design choice, or should we change them to non-repeated and require multiple mappings instead?

I believe this is a sound design that avoids needless repetition, but I'm open to counter examples in which a different design could be better.

Reply:
OK, so it does match the combination of all databases+schemas.

> I believe this is a sound design that avoids needless repetition, but I'm open to counter examples in which a different design could be better.

👍